src/caffe/solver.cpp

```diff
@@ -227,13 +210,44 @@ void Solver<Dtype>::Solve(const char* resume_file) {
         }
       }
     }
-
     ComputeUpdateValue();
     net_->Update();
+
+    // Save a snapshot if needed.
+    if (param_.snapshot() && (iter_ + 1) % param_.snapshot() == 0) {
+      Snapshot();
+    }
+  }
+}
+
+template <typename Dtype>
+void Solver<Dtype>::PreSolve() {
```

Comment:
Actually, we can't do that, because virtual functions can't be called from constructors. Here's a slightly more radical idea, which I've implemented now: get rid of `Solver::PreSolve` and just use the constructors, since constructors now provide the same functionality. `SGDSolver::PreSolve`, which was the only actual `PreSolve`, remains as it was and gets called from the `SGDSolver` constructor.